sql+cli: cannot dump UUID columns

SQL Dump does not seem to support UUID column types.

Steps to reproduce:

```
CREATE DATABASE test;
CREATE TABLE test.uuid (id UUID NOT NULL);
INSERT INTO test.uuid values ('e9716c74-2638-443d-90ed-ffde7bea7d1d');
```

Then running `./cockroach dump test --insecure` results in the following error:

```
# Server version: CockroachDB CCL v1.1-beta.20170907 (linux amd64, built 2017/09/13 03:03:08, go1.8.3) (same version as client)
# Cluster ID: 4628dcf5-e6e9-47a7-9b54-aef396723c9f
CREATE TABLE "uuid" (
	id UUID NOT NULL,
	FAMILY "primary" (id, rowid)
);
E170915 00:28:38.448152 1 util/log/crash_reporting.go:114  a panic has occurred!
panic: unknown []byte type: e9716c74-2638-443d-90ed-ffde7bea7d1d, id: UUID [recovered]
	panic: unknown []byte type: e9716c74-2638-443d-90ed-ffde7bea7d1d, id: UUID

goroutine 1 [running]:
github.com/cockroachdb/cockroach/pkg/util/log.RecoverAndReportPanic(0x7f9412274438, 0xc420014460, 0xc420442000)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:91 +0x9f
panic(0x1aed700, 0xc42026ce40)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/cli.dumpTableData(0x2a305e0, 0xc42000e018, 0xc4206f6700, 0xc42026c460, 0x1e, 0x3b, 0xc4206fe400, 0x1, 0x1caa966, 0x5, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/dump.go:473 +0x123c
github.com/cockroachdb/cockroach/pkg/cli.runDump(0x2a19c80, 0xc42025f7a0, 0x1, 0x2, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/dump.go:113 +0xa3a
github.com/cockroachdb/cockroach/pkg/cli.MaybeDecorateGRPCError.func1(0x2a19c80, 0xc42025f7a0, 0x1, 0x2, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/error.go:35 +0x5a
github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).execute(0x2a19c80, 0xc42025f6c0, 0x2, 0x2, 0x2a19c80, 0xc42025f6c0)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:644 +0x3f8
github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x2a17ac0, 0x6c0f9e, 0xc4202ab750, 0xc4206a1f10)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:734 +0x339
github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).Execute(0x2a17ac0, 0xc4206a1f00, 0xc4202ab720)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:693 +0x2b
github.com/cockroachdb/cockroach/pkg/cli.Run(0xc4200101d0, 0x3, 0x3, 0xc420014460, 0xc420442000)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:141 +0x6d
github.com/cockroachdb/cockroach/pkg/cli.Main()
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:50 +0x18d
main.main()
	/go/src/github.com/cockroachdb/cockroach/main.go:26 +0x20
```

Nothing special in the logs:

```
I170915 00:37:32.773638 363 sql/executor.go:715  [client=127.0.0.1:38074,user=root,n1] execRequest: SELECT rowid, id FROM test."uuid" AS OF SYSTEM TIME '1505435852759599189.0000000000' ORDER BY PRIMARY KEY test."uuid" LIMIT 10000
I170915 00:37:32.773752 363 sql/executor.go:1038  [client=127.0.0.1:38074,user=root,n1] executing 1/1: SELECT rowid, id FROM test."uuid" AS OF SYSTEM TIME '1505435852759599189.0000000000' ORDER BY PRIMARY KEY test."uuid" LIMIT 10000
I170915 00:37:32.773767 363 sql/executor.go:1353  [client=127.0.0.1:38074,user=root,n1] SELECT rowid, id FROM test."uuid" AS OF SYSTEM TIME '1505435852759599189.0000000000' ORDER BY PRIMARY KEY test."uuid" LIMIT 10000
I170915 00:37:32.775483 363 sql/executor.go:1852  [client=127.0.0.1:38074,user=root,n1] not distributing query
I170915 00:37:32.776033 363 sql/executor.go:1353  [client=127.0.0.1:38074,user=root,n1] SELECT (1 results) done
```

--- a/pkg/cli/dump.go
+++ b/pkg/cli/dump.go
@@ -420,6 +420,11 @@ func dumpTableData(w io.Writer, conn *sqlConn, clusterTS string, md tableMetadat
 						}
 					case "BYTES":
 						d = parser.NewDBytes(parser.DBytes(t))
+					case "UUID":
+						d, err = parser.ParseDUuidFromString(string(t))
+						if err != nil {
+							return err
+						}
 					default:
 						// STRING and DECIMAL types can have optional length
 						// suffixes, so only examine the prefix of the type.

--- a/pkg/cli/dump_test.go
+++ b/pkg/cli/dump_test.go
@@ -35,6 +35,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/cockroach/pkg/util/uuid"
 )
 
 func TestDumpRow(t *testing.T) {
@@ -55,11 +56,12 @@ func TestDumpRow(t *testing.T) {
 		n interval,
 		o bool,
 		e decimal,
+		u uuid,
 		tz timestamptz,
 		e1 decimal(2),
 		e2 decimal(2, 1),
 		s1 string(1),
-		FAMILY "primary" (i, f, d, t, n, o, tz, e1, e2, s1, rowid),
+		FAMILY "primary" (i, f, d, t, n, o, u, tz, e1, e2, s1, rowid),
 		FAMILY fam_1_s (s),
 		FAMILY fam_2_b (b),
 		FAMILY fam_3_e (e)
@@ -72,6 +74,7 @@ func TestDumpRow(t *testing.T) {
 		'2h30m30s',
 		true,
 		1.2345,
+		'e9716c74-2638-443d-90ed-ffde7bea7d1d',
 		'2016-01-25 10:10:10',
 		3.4,
 		4.5,
@@ -105,22 +108,23 @@ CREATE TABLE t (
 	n INTERVAL NULL,
 	o BOOL NULL,
 	e DECIMAL NULL,
+	u UUID NULL,
 	tz TIMESTAMP WITH TIME ZONE NULL,
 	e1 DECIMAL(2) NULL,
 	e2 DECIMAL(2,1) NULL,
 	s1 STRING(1) NULL,
-	FAMILY "primary" (i, f, d, t, n, o, tz, e1, e2, s1, rowid),
+	FAMILY "primary" (i, f, d, t, n, o, u, tz, e1, e2, s1, rowid),
 	FAMILY fam_1_s (s),
 	FAMILY fam_2_b (b),
 	FAMILY fam_3_e (e)
 );
 
-INSERT INTO t (i, f, s, b, d, t, n, o, e, tz, e1, e2, s1) VALUES
-	(1, 2.3, 'striiing', b'a1b2c3', '2016-03-26', '2016-01-25 10:10:10+00:00', '2h30m30s', true, 1.2345, '2016-01-25 10:10:10+00:00', 3, 4.5, 's'),
-	(NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL),
-	(NULL, '+Inf', NULL, NULL, NULL, NULL, NULL, NULL, 'Infinity', NULL, NULL, NULL, NULL),
-	(NULL, '-Inf', NULL, NULL, NULL, NULL, NULL, NULL, '-Infinity', NULL, NULL, NULL, NULL),
-	(NULL, 'NaN', NULL, NULL, NULL, NULL, NULL, NULL, 'NaN', NULL, NULL, NULL, NULL);
+INSERT INTO t (i, f, s, b, d, t, n, o, e, u, tz, e1, e2, s1) VALUES
+	(1, 2.3, 'striiing', b'a1b2c3', '2016-03-26', '2016-01-25 10:10:10+00:00', '2h30m30s', true, 1.2345, 'e9716c74-2638-443d-90ed-ffde7bea7d1d', '2016-01-25 10:10:10+00:00', 3, 4.5, 's'),
+	(NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL),
+	(NULL, '+Inf', NULL, NULL, NULL, NULL, NULL, NULL, 'Infinity', NULL, NULL, NULL, NULL, NULL),
+	(NULL, '-Inf', NULL, NULL, NULL, NULL, NULL, NULL, '-Infinity', NULL, NULL, NULL, NULL, NULL),
+	(NULL, 'NaN', NULL, NULL, NULL, NULL, NULL, NULL, 'NaN', NULL, NULL, NULL, NULL, NULL);
 `
 
 	if string(out) != expect {
@@ -352,7 +356,8 @@ func TestDumpRandom(t *testing.T) {
 			e decimal,
 			s string,
 			b bytes,
-			PRIMARY KEY (rowid, i, f, d, m, n, o, e, s, b)
+			u uuid,
+			PRIMARY KEY (rowid, i, f, d, m, n, o, e, s, b, u)
 		);
 	`, nil); err != nil {
 		t.Fatal(err)
@@ -402,6 +407,15 @@ func TestDumpRandom(t *testing.T) {
 				t.Fatal(err)
 			}
 
+			uuidBytes := make([]byte, 16)
+			if _, err := rnd.Read(b); err != nil {
+				t.Fatal(err)
+			}
+			u, err := uuid.FromBytes(uuidBytes)
+			if err != nil {
+				t.Fatal(err)
+			}
+
 			vals := []driver.Value{
 				_i,
 				i,
@@ -413,15 +427,16 @@ func TestDumpRandom(t *testing.T) {
 				[]byte(e), // decimals come out as `[]byte`s
 				string(s),
 				b,
+				[]byte(u.String()),
 			}
-			if err := conn.Exec("INSERT INTO d.t VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)", vals); err != nil {
+			if err := conn.Exec("INSERT INTO d.t VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)", vals); err != nil {
 				t.Fatal(err)
 			}
 			generatedRows = append(generatedRows, vals[1:])
 		}
 
 		check := func(table string) {
-			q := fmt.Sprintf("SELECT i, f, d, m, n, o, e, s, b FROM %s ORDER BY rowid", table)
+			q := fmt.Sprintf("SELECT i, f, d, m, n, o, e, s, b, u FROM %s ORDER BY rowid", table)
 			nrows, err := conn.Query(q, nil)
 			if err != nil {
 				t.Fatal(err)

--- a/pkg/sql/copy_in_test.go
+++ b/pkg/sql/copy_in_test.go
@@ -53,6 +53,7 @@ func TestCopyNullInfNaN(t *testing.T) {
 			n INTERVAL NULL,
 			o BOOL NULL,
 			e DECIMAL NULL,
+			u UUID NULL,
 			tz TIMESTAMP WITH TIME ZONE NULL
 		);
 	`); err != nil {
@@ -70,10 +71,10 @@ func TestCopyNullInfNaN(t *testing.T) {
 	}
 
 	input := [][]interface{}{
-		{nil, nil, nil, nil, nil, nil, nil, nil, nil, nil},
-		{nil, math.Inf(1), nil, nil, nil, nil, nil, nil, nil, nil},
-		{nil, math.Inf(-1), nil, nil, nil, nil, nil, nil, nil, nil},
-		{nil, math.NaN(), nil, nil, nil, nil, nil, nil, nil, nil},
+		{nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil},
+		{nil, math.Inf(1), nil, nil, nil, nil, nil, nil, nil, nil, nil},
+		{nil, math.Inf(-1), nil, nil, nil, nil, nil, nil, nil, nil, nil},
+		{nil, math.NaN(), nil, nil, nil, nil, nil, nil, nil, nil, nil},
 	}
 
 	for _, in := range input {
@@ -143,6 +144,7 @@ func TestCopyRandom(t *testing.T) {
 			t TIMESTAMP,
 			s STRING,
 			b BYTES,
+			u UUID,
 			tz TIMESTAMP WITH TIME ZONE
 		);
 	`); err != nil {
@@ -154,7 +156,7 @@ func TestCopyRandom(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	stmt, err := txn.Prepare(pq.CopyInSchema("d", "t", "id", "n", "o", "i", "f", "e", "t", "s", "b", "tz"))
+	stmt, err := txn.Prepare(pq.CopyInSchema("d", "t", "id", "n", "o", "i", "f", "e", "t", "s", "b", "u", "tz"))
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -168,6 +170,7 @@ func TestCopyRandom(t *testing.T) {
 		sqlbase.ColumnType_TIMESTAMP,
 		sqlbase.ColumnType_STRING,
 		sqlbase.ColumnType_BYTES,
+		sqlbase.ColumnType_UUID,
 		sqlbase.ColumnType_TIMESTAMPTZ,
 	}
 

--- a/pkg/sql/parser/col_types_test.go
+++ b/pkg/sql/parser/col_types_test.go
@@ -48,6 +48,7 @@ func TestParseColumnType(t *testing.T) {
 		{"NUMERIC", &DecimalColType{Name: "NUMERIC"}},
 		{"NUMERIC(8)", &DecimalColType{Name: "NUMERIC", Prec: 8}},
 		{"NUMERIC(9,10)", &DecimalColType{Name: "NUMERIC", Prec: 9, Scale: 10}},
+		{"UUID", &UUIDColType{}},
 		{"DATE", &DateColType{}},
 		{"TIMESTAMP", &TimestampColType{}},
 		{"TIMESTAMP WITH TIME ZONE", &TimestampTZColType{}},

--- a/pkg/sql/parser/datum.go
+++ b/pkg/sql/parser/datum.go
@@ -1085,7 +1085,13 @@ func (*DUuid) AmbiguousFormat() bool { return false }
 
 // Format implements the NodeFormatter interface.
 func (d *DUuid) Format(buf *bytes.Buffer, f FmtFlags) {
-	encodeSQLString(buf, d.UUID.String())
+	if !f.bareStrings {
+		buf.WriteByte('\'')
+	}
+	buf.WriteString(d.UUID.String())
+	if !f.bareStrings {
+		buf.WriteByte('\'')
+	}
 }
 
 // Size implements the Datum interface.

